ProcessManager.py

```diff
@@ -1,88 +1,136 @@
+    Manages processes that run in the robocluster framework.
+
+    Processes are just programs that can run independantly
+    from each other, in any language supported by the robocluster library.
+    The ProcessManager is in charge of starting and stopping processes,
+    and monitering their status and handle the event of a crash
+    or a process not responding.
     """
-    
-    def __init__(self, **kwargs):
-        """Initializes the process manager"""
-        self.T = [] # Thread or process library (It is right now a list but it could be changed to something else other than a list).
-        return super().__init__(**kwargs)
+
+    def __init__(self):
+        """Initialize a process manager."""
+        # process dictionary stores RoboProcess instances indexed
+        # by the process name
+        self.processes = {}
+
+    def __enter__(self):
+        """Enter context manager."""
+        return self
+
+    def __exit__(self, *exc):
+        """Exit context manager, makes sure all processes are stopped."""
+        self.stopAllProcesses()
+        return False
 
     def isEmpty(self):
-        if T == []:
-            return True
-        else:
-            return False
+        """Return if processes is empty."""
+        return len(self.processes) == 0
 
-    def createProcess(self):
+    def createProcess(self, name, command):
         """
-        TODO create a function that takes in parameters and uses those parameters to
-        ceate one single process. The process must then be stored in "self.T".
+        Create a process.
+
+        Arguments:
+        name    - name to identify process, must be unique to process manager.
+        command - shell command for process to execute.
         """
+        if name in self.processes:
+            raise ValueError(f'Process with the same name exists: {name}')
+
+        if not isinstance(command, str):
+            raise ValueError('command must be a string')
 
+        self.processes[name] = RoboProcess(command)
+
+    # TODO: Do we really need this if we have createProcess?
     def createThread(self):
         """
         TODO create a function that takes in parameters and uses those parameters to
         ceate one single thread. The thread must then be stored in "self.T".
         """
+        pass
+
+    def startProcess(self, name):
+        """Start a single process."""
+        print('Starting {}'.format(name))
+        self.processes[name].execute()
 
     def startAllProcesses(self):
-        """
-        TODO: create a function that starts all processes or threads in "self.T" all at once.
-        """
+        """Start all managed processes."""
+        for process in self.processes:
+            self.startProcess(process)
+
+    def stopProcess(self, name):
+        """Stop a process by name."""
+        print ("Shutting down " + name)
+        self.processes[name].popen.kill()
+        self.processes[name].popen.wait()
 
     def stopAllProcesses(self):
```

Comment:
I think having this syntax `manager.stop()` would be better. We could also use an optional `name` kwarg to stop a single process.